nixos/keycloak: Init

talyz opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
Add a module for [Keycloak](https://www.keycloak.org/), an open source identity and access management server.

I just found out about #91734 - my intention was not to step on your toes, @ngerstle; I started work on this module in early 2019 and have had it deployed at work since then, but just now had time to update and polish it for upstreaming.

The module features explicit configuration options for common settings put into the `standalone.xml` configuration file (address, port, tls, etc) and allows for arbitrary settings through an `extraConfig` option. The settings are applied on top of the shipped default configuration via the jboss-cli.

Fixes #87673.

###### Todo

- [x] Write test(s)
- [x] Document and clean up the jboss-cli functions
- [x] Add a section to the manual

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

talyz committed:
nixos/keycloak: Init

bbigras commented:
> just found out about #99517

did you mean to link #91734?

talyz commented:
@bbigras I did indeed. Thanks!

aanderse commented:
I might be recalling incorrectly, but I thought @Mic92 had some comments about relying on ACLs in our modules.

talyz commented:
@aanderse Aha, well, I don't think the `setfacl` commands are actually crucial here - they were mostly added as a safeguard IIRC.

aanderse commented on nixos/modules/services/web-apps/keycloak.nix:
```diff
@@ -0,0 +1,441 @@
+          script = ''
+            set -eu
+
+            if ! [[ -e ${cfg.databasePasswordFile} ]]; then
```
I would prefer declarative [ConditionPathExists](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Conditions%20and%20Asserts) over imperative `bash` statements.

aanderse commented on nixos/modules/services/web-apps/keycloak.nix:
```diff
@@ -0,0 +1,441 @@
+                  cp -r ${cfg.package}/standalone/configuration /var/lib/keycloak/
+              fi
+
+              if ! [[ -e ${cfg.databasePasswordFile} ]]; then
```
As above, I would prefer declarative [ConditionPathExists](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Conditions%20and%20Asserts) over imperative `bash` statements.

talyz replied:
I changed this for the `keycloak` unit, but I'm not sure that it's an improvement - instead of getting an error when the file doesn't exist you now get a silent error that looks like a success.

talyz replied:
Reverted to checking with `bash`, since it makes runtime errors more explicit.

aanderse replied:
@talyz unfortunate, but that makes sense to me. I'll ping some `systemd` keeners like @andi, @flokli, @arianvp, etc... just to make sure they don't have any feedback I've missed on `ConditionPathExists`.

talyz replied:
I found a solution - for every `ConditionX` there's also an `AssertX` which fails the unit instead of skipping it. It still doesn't give great error output: it only logs `keycloak.service: Starting requested but asserts failed.`, but if you check the status of the unit you can see the failed condition, so still much better than just skipping it.

arianvp replied:
Why do we need to assert in the first place?  Doesn't keycloak already crash with a happy error message if these files are missing?

talyz replied:
Well, the initial reason for adding them were to get a better error message than
```
/nix/store/ad05hhpmg2mhn6jqpdgpn1c8xllqdgmk-unit-script-keycloakDatabaseInit-start/bin/keycloakDatabaseInit-start: line 6: /run/keys/db_password: No such file or directory
```
but I suppose that might actually be less cryptic than the assert message.

talyz replied:
I removed the asserts. If any of the files are configured incorrectly keycloak won't be started and the error message will be pretty clear.

aanderse commented on nixos/modules/services/web-apps/keycloak.nix:
```diff
@@ -0,0 +1,441 @@
+              db_password="$(<'${cfg.databasePasswordFile}')"
+              ${pkgs.replace}/bin/replace-literal -fe '@db-password@' "$db_password" /var/lib/keycloak/configuration/standalone.xml
+
+              mkdir -p /var/lib/keycloak/{deployments,ssl}
```
I would prefer declarative [StateDirectory](https://www.freedesktop.org/software/systemd/man/systemd.exec.html) over imperative `bash` statements.

talyz replied:
I rewrote this; it now uses `RuntimeDirectory` for all required directories.

aanderse commented on nixos/modules/services/web-apps/keycloak.nix:
```diff
@@ -0,0 +1,441 @@
+              chmod -R u+rwX,g-rwx,o-rwx /var/lib/keycloak/
+              ${pkgs.acl}/bin/setfacl -R -m d:u:keycloak:rwX /var/lib/keycloak/
+              ${pkgs.acl}/bin/setfacl -R -m d:g::-,d:o::- /var/lib/keycloak/
+              chown keycloak:keycloak /run/wrappers/bin/keycloak
```
Possibly it is obvious... but I'm not following the purpose of this.

talyz replied:
Well, it was a quick and dirty way to assign the correct permissions for current and future files and the wrapper was needed when I, for some reason, decided to use `CapabilityBoundingSet` instead of `AmbientCapabilities`. All of it is removed now :)

aanderse commented on nixos/modules/services/web-apps/keycloak.nix:
```diff
@@ -0,0 +1,441 @@
+            if [[ "$current_owner" != "${cfg.databaseUsername}" ]]; then
+                $PSQL -tAc 'ALTER DATABASE "keycloak" OWNER TO "${cfg.databaseUsername}"'
+                if [[ -e "${config.services.postgresql.dataDir}/.reassigning_keycloak" ]]; then
+                    echo "Reassigning ownership of database keycloak to user ${cfg.databaseUsername} failed on last boot. Failing..." >&2
```
This is scary to me. I won't make a big deal about it, but I *personally* don't think we should reassign sole ownership databases this way.

talyz replied:
I've removed this this and now force the local db username to be `keycloak` - I figure there isn't really a need for it to be configurable when a local db is provisioned automatically.

aanderse commented on nixos/modules/services/web-apps/keycloak.nix:
```diff
@@ -0,0 +1,441 @@
+            JBOSS_MODULEPATH = "${cfg.package}/modules";
+          };
+          serviceConfig = {
+            ExecStartPre = systemdPrefixScript "startPre" "+" (''
```
What are the bits that require this to be run as `root`? I feel like we might be able to get around this... :thinking:

talyz replied:
Yeah, I'm working on this right now. I _think_ the only part that needs root access is the ownership reassignment and permission changes to files, but I'll narrow it down to as few commands as possible and split the script into a privileged and an unprivileged part.

talyz replied:
It now uses root access only to get the secrets, which is a trade-off that afaik has to be made when `DynamicUser` is set.

arianvp replied:
Correct.  However this will be fixed in future Systemd-versions  https://github.com/systemd/systemd/blob/master/NEWS#L320

aanderse reviewed (commented):
I have left some discussion points I hope aren't entirely useless.

talyz committed:
nixos/keycloak: Add test

talyz committed:
nixos/keycloak: Document internal functions

ngerstle commented:
> ###### Motivation for this change
> 
> Add a module for [Keycloak](https://www.keycloak.org/), an open source identity and access management server.
> 
> I just found out about #91734 - my intention was not to step on your toes, @ngerstle; I started work on this module in early 2019 and have had it deployed at work since then, but just now had time to update and polish it for upstreaming.
> 
> The module features explicit configuration options for common settings put into the `standalone.xml` configuration file (address, port, tls, etc) and allows for arbitrary settings through an `extraConfig` option. The settings are applied on top of the shipped default configuration via the jboss-cli.
> 
> Fixes #87673.


No worries, @talyz - this is far beyond what I was slowly working on!

talyz committed:
nixos/keycloak: Add documentation

talyz committed:
rl-2103: Note the addition of the Keycloak service

talyz committed:
nixos/keycloak: Add support for MySQL and external DBs with SSL

- Add support for using MySQL as an option to PostgreSQL.
- Enable connecting to external DBs with SSL
- Add a database port config option

talyz commented:
I've added support for `MySQL` and secure external database connections, which was the last thing I wanted to get in. I consider this done now and will merge in a few days unless someone objects to it.
